tools/yul_phaser/Chromosome.cpp

```diff
@@ -0,0 +1,190 @@
+#include <libyul/optimiser/ForLoopInitRewriter.h>
+#include <libyul/optimiser/FullInliner.h>
+#include <libyul/optimiser/FunctionGrouper.h>
+#include <libyul/optimiser/FunctionHoister.h>
+#include <libyul/optimiser/LoadResolver.h>
+#include <libyul/optimiser/LoopInvariantCodeMotion.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/RedundantAssignEliminator.h>
+#include <libyul/optimiser/Rematerialiser.h>
+#include <libyul/optimiser/SSAReverser.h>
+#include <libyul/optimiser/SSATransform.h>
+#include <libyul/optimiser/StructuralSimplifier.h>
+#include <libyul/optimiser/Suite.h>
+#include <libyul/optimiser/UnusedPruner.h>
+#include <libyul/optimiser/VarDeclInitializer.h>
+#include <libyul/optimiser/VarNameCleaner.h>
+
+#include <cassert>
+#include <sstream>
+
+using namespace std;
+using namespace solidity::yul;
+
+namespace solidity::phaser
+{
+
+Chromosome Chromosome::make_random(int _length)
+{
+	assert(_length >= 0);
+
+	vector<string> steps;
+	steps.reserve(_length);
+	for (auto i = 0; i < _length; ++i)
+		steps.push_back(randomOptimizationStep());
+
+	return Chromosome(move(steps));
+}
+
+ostream& operator<<(ostream& _stream, Chromosome const& _chromosome)
+{
+	for (auto stepName: _chromosome.m_optimizationSteps)
+		_stream << Chromosome::abbreviateStepName(stepName);
+
+	return _stream;
+}
+
+char Chromosome::abbreviateStepName(string const& _stepName)
+{
+	static unordered_map<string, char> stepToSymbol{
+		{BlockFlattener::name,                'f'},
+		{CommonSubexpressionEliminator::name, 'c'},
+		{ConditionalSimplifier::name,         'C'},
+		{ConditionalUnsimplifier::name,       'U'},
+		{ControlFlowSimplifier::name,         'n'},
+		{DeadCodeEliminator::name,            'D'},
+		{EquivalentFunctionCombiner::name,    'v'},
+		{ExpressionInliner::name,             'e'},
+		{ExpressionJoiner::name,              'j'},
+		{ExpressionSimplifier::name,          's'},
+		{ExpressionSplitter::name,            'x'},
+		{ForLoopConditionIntoBody::name,      'I'},
+		{ForLoopConditionOutOfBody::name,     'O'},
+		{ForLoopInitRewriter::name,           'o'},
+		{FullInliner::name,                   'i'},
+		{FunctionGrouper::name,               'g'},
+		{FunctionHoister::name,               'h'},
+		{LiteralRematerialiser::name,         'T'},
+		{LoadResolver::name,                  'L'},
+		{LoopInvariantCodeMotion::name,       'M'},
+		{RedundantAssignEliminator::name,     'r'},
+		{Rematerialiser::name,                'm'},
+		{SSAReverser::name,                   'V'},
+		{SSATransform::name,                  'a'},
+		{StructuralSimplifier::name,          't'},
+		{UnusedPruner::name,                  'u'},
+		{VarDeclInitializer::name,            'd'},
+		{VarNameCleaner::name,                'l'},
+	};
+
+	auto pair = stepToSymbol.find(_stepName);
+	assert(pair != stepToSymbol.end());
+
+	return pair->second;
+}
+
+vector<string> Chromosome::allStepNames()
+{
+	assert(not OptimiserSuite::allSteps().empty());
```

Comment:
`not`? :)

Reply:
You mean why I didn't use `!`? I think that `not`, `and` and `or` read better than `!`, `&&` and `||`. But that's just my personal preference. I can switch if you prefer to just have `!`, `&&`, `||` consistently everywhere.

BTW this assertion is no longer necessary. I changed the code here and forgot to remove it. Originally `stepNames` below was a `static` variable initialized the first time the function runs. If it's empty, it means that it has not been initialized yet. But that's not necessarily true if `OptimiserSuite::allSteps()` for whatever reason happens to be empty too. That probably won't happen under normal circumstances and even it did, it won't break the code, just make it unnecessarily reinitialize the variable but I wanted to make this corner case clear.

Reply:
Looks like MSVC 2017 in AppVeyor does not like it :(
Replaced with `!`, `&&` and `||`.